sys/linux/fs_ioctl_btrfs.txt

```diff
@@ -1,7 +1,455 @@
+ioctl$BTRFS_IOC_QUOTA_RESCAN(fd fd, cmd const[BTRFS_IOC_QUOTA_RESCAN], arg ptr[in, btrfs_ioctl_quota_rescan_args])
+ioctl$BTRFS_IOC_QUOTA_RESCAN_STATUS(fd fd, cmd const[BTRFS_IOC_QUOTA_RESCAN_STATUS], arg ptr[out, btrfs_ioctl_quota_rescan_args])
+ioctl$BTRFS_IOC_QUOTA_RESCAN_WAIT(fd fd, cmd const[BTRFS_IOC_QUOTA_RESCAN_WAIT], arg const[0])
+ioctl$BTRFS_IOC_DEV_REPLACE(fd fd, cmd const[BTRFS_IOC_DEV_REPLACE], arg ptr[inout, btrfs_ioctl_dev_replace_args])
+ioctl$BTRFS_IOC_GET_SUPPORTED_FEATURES(fd fd, cmd const[BTRFS_IOC_GET_SUPPORTED_FEATURES], arg ptr[out, array[btrfs_ioctl_feature_flags, 3]])
+ioctl$BTRFS_IOC_GET_FEATURES(fd fd, cmd const[BTRFS_IOC_GET_FEATURES], arg ptr[out, btrfs_ioctl_feature_flags])
+ioctl$BTRFS_IOC_SET_FEATURES(fd fd, cmd const[BTRFS_IOC_SET_FEATURES], arg ptr[in, btrfs_ioctl_feature_flags])
+ioctl$BTRFS_IOC_GET_SUBVOL_INFO(fd fd, cmd const[BTRFS_IOC_GET_SUBVOL_INFO], arg ptr[out, btrfs_ioctl_get_subvol_info_args])
+ioctl$BTRFS_IOC_GET_SUBVOL_ROOTREF(fd fd, cmd const[BTRFS_IOC_GET_SUBVOL_ROOTREF], arg ptr[inout, btrfs_ioctl_get_subvol_rootref_args])
+ioctl$BTRFS_IOC_INO_LOOKUP_USER(fd fd, cmd const[BTRFS_IOC_INO_LOOKUP_USER], arg ptr[inout, btrfs_ioctl_ino_lookup_user_args])
+
+resource transid[int64]
+resource devid[int64]
+resource treeid[int64]
+resource dirid[int64]
+# These should be a resource but it's not returned by any syscall. So we temporarily describe them as type
+type qgroupid int64
+type objectid int64
+type subvolid objectid
+#NEED: uuid should be a resource
+
+define BTRFS_PATH_MAX	BTRFS_PATH_NAME_MAX+1 
+define BTRFS_SUBVOL_MAX	BTRFS_SUBVOL_NAME_MAX+1
+
+#The following two struct needs to be customized to fit the need of each specific ioctls calling them
+#But I still need more information in order to make the change
+btrfs_ioctl_vol_args {
+	fd	align64[fd]
+	name	array[int8, BTRFS_PATH_MAX]
+}
+
+btrfs_ioctl_vol_args_v2 {
+	fd	align64[fd]
+	transid	transid
+	flags	flags[btrfs_ioctl_vol_args_v2_flags, int64]
+	u1	btrfs_vol_args_v2_u1
+	u2	btrfs_vol_args_v2_u2
+}
+
+btrfs_vol_args_v2_u1 [
+	inherit	btrfs_vol_args_v2_u1_s1
+	unused	array[int64, 4]
+]
+
+btrfs_vol_args_v2_u2 [
+	name		array[int8, BTRFS_SUBVOL_MAX]
+	devid		devid
+	subvolid	subvolid
+]
+
+btrfs_vol_args_v2_u1_s1 {
+	size		bytesize[qgroup_inherit, int64]
+	qgroup_inherit	ptr[in, btrfs_qgroup_inherit]
+}
+
+btrfs_qgroup_inherit {
+	flags		flags[btrfs_qgroup_inherit_flags, int64]
+	num_qgroups	len[qgroups, int64]
+	num_ref_copies	int64
+	num_excl_copies	int64
+	lim		btrfs_qgroup_limit
+	qgroups		array[qgroupid]
+}
+
+btrfs_qgroup_limit {
+	flags		flags[btrfs_qgroup_limit_flags, int64]
+	max_rfer	int64
+	max_excl	int64
+	rsv_rfer	int64
+	rsv_excl	int64
+}
+
+btrfs_ioctl_defrag_range_args {
+	start		int64
+	len		int64
+	flags		flags[btrfs_ioctl_defrag_range_args_flags, int64]
+	extent_thresh	int32
+	compress_type	flags[btrfs_compression_type, int32]
+	unused		array[int32, 4]
+}
+
+btrfs_ioctl_fs_info_args {
```

Comment:
There are a lot of differences between this and the declaration in the kernel source tree.  Did you look in the right place?  The header has:

	#define BTRFS_FS_INFO_FLAG_CSUM_INFO			(1 << 0)
	#define BTRFS_FS_INFO_FLAG_GENERATION			(1 << 1)
	#define BTRFS_FS_INFO_FLAG_METADATA_UUID		(1 << 2)

	struct btrfs_ioctl_fs_info_args {
		__u64 max_id;				/* out */
		__u64 num_devices;			/* out */
		__u8 fsid[BTRFS_FSID_SIZE];		/* out */
		__u32 nodesize;				/* out */
		__u32 sectorsize;			/* out */
		__u32 clone_alignment;			/* out */
		/* See BTRFS_FS_INFO_FLAG_* */
		__u16 csum_type;			/* out */
		__u16 csum_size;			/* out */
		__u64 flags;				/* in/out */
		__u64 generation;			/* out */
		__u8 metadata_uuid[BTRFS_FSID_SIZE];	/* out */
		__u8 reserved[944];			/* pad to 1k */
	};

Reply:
Hmm I was looking at Linux v5.8 ("https://elixir.bootlin.com/linux/v5.8/source/include/uapi/linux/btrfs.h"). Should I change them to follow the source?

Reply:
Oh, I see this was just changed in 5.9.

I suggest you just mention in the commit message that you're describing the API as of Linux v5.8, so that it isn't a moving target.

Reply:
Sounds good!